README
Overview
========

The aim is to have an editor in which content is edited semantically, and both
column layout and styling are applied separately.  However, in the database
only the combined HTML is stored.  So, we have the following requirements:

HTML content is stored in the database something like:

 <div class="row columns2">
   <div class="column">
     <h1 class="fancy">Heading</h1>
     <p class="note bordered">Some text</p>
   </div>
   <div class="column">
     <p>Some more text</p>
   </div>
 </div>

This is loaded into a custom editor in the browser which then immediately does
some AJAX calls to the server to decompose into simplified HTML:

 <h1 id="h1_1">Heading</h1>
 <p id="p_1">Some text</p>
 <p id='p_2'>Some more text</p>

 and an array of objects specifying presentation e.g.

 [
   {'h1_1': ['fancy']},                # styles for the H1
   {'p_1':  ['note', 'bordered']},     # styles for the P
   {'newrow_h1_1': 'newrow' },   # specifies row before h1_1
   {'newrow_p_2': 'newcolum' },
 ]

These two parts are then edited separately, i.e. the user does not see the
combined HTML while editing.  Before saving to the database, the editor does an
AJAX call to combine the two parts.  Note the use of the 'id' attributes are
added to help identify what the styling information belongs to -- these will be
removed before saving in the database.

Depending on the GUI for editing the presentation info, the user may have to
press a 'refresh' button so that they can assign presentation info to newly
entered paragraphs and headings etc.  Newly entered paragraphs and headings will
obviously not have the 'id' attributes, so the HTML may have to be updated at
this point as well, so everything has an id.


WYMeditor extensions
====================

The semantic editor application provides its own version of WYMeditor.  This
is derived from the skins/templates found in django-cms2, with the following
customisations:

 - name of the skin changed from 'django' to 'semanticeditor'.  This is
   necessary if we are to allow two different versions of WYMeditor on the
   same page, one using the normal WYMeditor + CMS controls, another with
   the full 'semantic editor' controls, for longer pieces of content.

 - The skin.js file removes the 'classes' panel, since this is not needed
   and is replaced by dynamic elements.

 - skins.css - added styling for additional presentation controls

 - a WYMeditor plugin named 'semantic' has been added.  This implements
   most of the client side logic for this application.

 - the editorwidget.html template has been changed.  In particular
   - the skin has been changed to 'semanticeditor'
   - a call to wymeditor.semantic() has been added to set up the plugin.

 - a new django widget 'SemanticEditor' has been created that inherits from
   the django 'WYMEditor' widget in django-cms2.  This is needed so that
    - the WYMeditor 'semantic' plugin javascript file can be added to Media
    - our editorwdiget.html template can be used.



CHANGES.rst
Version 0.3.1
-------------
* Fixed issue with CssClass.templates field which could become corrupt via a
  roundtrip through JSON. (It ended up with things like "u['']" etc.)  If this
  affects you, you will need to manual check this field and correct it.

Version 0.3
-----------
* Compatibility with django-cms 2.3 and 2.4
* Grouping of CSS classes by category
* Button to open editor in new tab
* UI improvements
* Fixed typo that caused complete failure of semantic plugin under jQuery 1.4.2
* Other bug fixes

Version 0.2.1
-------------

* Fixed packaging bug (no templates or static media)

Version 0.2
-----------

* 'inner row' and 'inner column' commands
* Added ability to group CSS classes by category
* Compatibility with Django 1.3 and django-cms 2.1.3
* Lots of bug fixes and refinements


Version 0.1
-----------

Initial release




semanticeditor/views.py
from cms.models import Page
from django.http import HttpResponse
from django.utils import simplejson
from django.core.mail import mail_admins
from django.conf import settings
from django.utils.translation import ugettext as _
from semanticeditor.api import extract_presentation, format_html, preview_html, AllUserErrors, COMMANDS, PresentationInfo, PresentationClass, clean_html, get_classes
from semanticeditor.models import CssClass
import sys
try:
    from functools import wraps
except ImportError:
    from django.utils.functional import wraps  # Python 2.3, 2.4 fallback.


def json_view(func):
    """
    Use this decorator on a function that takes a request and returns
    a dictionary of values in order to create a view that handles
    errors and return JSON.

    The dictionary should be in this standard format:

    {'result': 'ok',
     'value': your_actual_value
    }

    or

    {'result': 'usererror',
     'message': an_error_message

    or

    {'result': 'error',
     'message': an_error_message
    }
    """
    def wrapper(request, *a, **kw):
        response = None
        try:
            response = func(request, *a, **kw)
        except Exception, e:
            # Mail the admins with the error
            exc_info = sys.exc_info()
            subject = 'JSON view error: %s' % request.path
            try:
                request_repr = repr(request)
            except:
                request_repr = 'Request repr() unavailable'
            import traceback
            message = 'Traceback:\n%s\n\nRequest:\n%s' % (
                '\n'.join(traceback.format_exception(*exc_info)),
                request_repr,
                )
            mail_admins(subject, message, fail_silently=True)

            # Come what may, we're returning JSON.
            if hasattr(e, 'message'):
                msg = e.message
            else:
                msg = _('Internal error')+': '+ str(e)
            response = error(msg)

        json = simplejson.dumps(response)
        return HttpResponse(json, mimetype='application/json')

    return wraps(func)(wrapper)


def error(msg):
    """
    Standard error result - for internal errors
    """
    return dict(result='error', message=msg)


def failure(msg):
    """
    Standard failure result
    """
    return dict(result='usererror', message=msg)


def success(value):
    """
    Standard success result
    """
    return dict(result='ok', value=value)


# Anything that depends on values the user may have entered and might
# contain 'errors' should use this, normally passing in AllUserErrors
# as 'exceptions'.  'server' errors are handled by @json_view
def graceful_errors(exceptions, callback):
    """
    Retrieve a value from a callback, handling the exceptions that
    are passed in, and returning in standard formats
    """
    try:
        val = callback()
    except exceptions, e:
        return failure(e.args[0])
    return success(val)


def PI_to_dict(pi):
    """
    Converts a PresentationInfo to a dictionary
    for use client side
    """
    d = {}
    # We need to filter out things that are server side only and can't be
    # turned into JSON.
    allowed = set([str, unicode, int, bool, dict, list])
    for k, v in pi.__dict__.items():
        if type(v) in allowed:
            d[k] = v
    return d


def dict_to_PI(d, classes):
    """
    Convert a dictionary to a PresentationInfo,
    using a pre-fetched dictionary of CssClass objects
    """
    if d['prestype'] == 'command':
        return PresentationInfo(prestype=d['prestype'], name=d['name'])
    else:
        c = classes.get(d['name'])
        if c is None:
            return None
        else:
            return css_class_to_presentation_class(c)


def css_class_to_presentation_class(c):
    return PresentationClass(c.name,
                             verbose_name=c.verbose_name,
                             description=c.description,
                             allowed_elements=c.allowed_elements.lower().split(' '),
                             column_equiv=c.column_equiv,
                             category=c.category.name if c.category is not None else None)


@json_view
def retrieve_styles(request):
    template = request.GET['template']
    page_id = request.GET['page_id']
    if template == settings.CMS_TEMPLATE_INHERITANCE_MAGIC:
        # Need to look up page to find out what template to use
        p = Page.objects.get(pk=page_id)
        template = p.get_template()
    classes = get_classes(template)
    retval = map(css_class_to_presentation_class,
                 classes)
    return success(map(PI_to_dict,retval))


@json_view
def retrieve_commands(request):
    return success(map(PI_to_dict, COMMANDS))


@json_view
def separate_presentation(request):
    """
    Returns a JSON object:
     { presentation: <dictionary of presentation info from html>
       html: <input html stripped of presentation>
     }
    """
    data = request.POST.get('html','')

    def _handled():
        pres, html = extract_presentation(data)
        # Rewrite pres so that we can serialise it to JSON
        pres2 = {}
        for k, v in pres.items():
            pres2[k] = [PI_to_dict(p) for p in v]
        return dict(presentation=pres2,
                    html=html)

    return graceful_errors(AllUserErrors, _handled)


def _convert_pres(pres):
    # Convert dictionaries into PresentationInfo classes. We need actual
    # CssClass instances in order to be able to restore column_equiv and
    # allowed_elements info
    classes = dict((c.name, c) for c in CssClass.objects.all())
    retval = {}
    for k, v in pres.items():
        # v is list of PI dicts
        newlist = []
        for i, item in enumerate(v):
            pi = dict_to_PI(item, classes)
            # if CssClass was removed from DB, pi can be None
            if pi is not None:
                newlist.append(pi)
        retval[k]= newlist
    return retval


@json_view
def combine_presentation(request):
    """
    Combines submitted 'html' and 'presentation' data,
    returning a dictionary containing { html: <combined html> }
    """
    html = request.POST.get('html', '')
    presentation = request.POST.get('presentation', '{}')
    presentation = simplejson.loads(presentation)
    presentation = _convert_pres(presentation)
    return graceful_errors(AllUserErrors, lambda: dict(html=format_html(html, presentation, pretty_print=True)))


@json_view
def preview(request):
    html = request.POST.get('html', '')
    presentation = request.POST.get('presentation', '{}')
    presentation = simplejson.loads(presentation)
    presentation = _convert_pres(presentation)

    return graceful_errors(AllUserErrors, lambda: dict(html=preview_html(html, presentation)))


@json_view
def clean_html_view(request):
    html = request.POST.get('html', '')
    return graceful_errors(AllUserErrors, lambda: dict(html=clean_html(html)))
